feat: write a detailed README for learning a policy in Isaac Lab

yrh012 opened the pull request:
<!-- Pull Request guidelines:
1. Give the PR a relevant and descriptive title, using one of the supported commit types:
   [ build, ci, chore, docs, feat, fix, perf, refactor, release, revert, style, test ]
2. Link the PR to the parent issue, which should already describe and motivate the PR
3. Explain how the PR addresses the parent issue
4. Add any supporting resources (screenshots, test results, etc)
5. Help the reviewer by suggesting the method and estimated time to review
6. If applicable, make a checklist of tasks that should be completed before merging
7. If applicable, mention related issues (blocked by / blocks)
8. Post creation actions: tag reviewers and link the PR to its parent issue under the Development section
-->

## Description

<!-- Required: link the parent issue -->

- #9   

<!-- Required: explain how the PR addresses the parent issue -->
This PR addresses the issue by providing a comprehensive README that explains how to set up a new reinforcement learning environment in Isaac Lab, highlights key concepts for new users, presents the available reinforcement learning libraries, demonstrates how to train a policy, and concludes with a final note on the reward-to-RL Policy Component SDK.

<!-- Optional: add additional resources (links, screenshots, test results, etc) -->

## Review guidelines

<!-- Required: estimate how long a review should take -->
Estimated Time of Review: 15 minutes

<!-- Optional: provide more suggestions such as "Review by commit", "Read this documentation first", etc -->

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

<!-- Optional: define further tasks that should be completed before merging
- [x] Task 1
- [ ] Task 2
-->

<!-- Optional: link related issues
## Related issues
### Blocked by:
- #0

### Blocks:
- #0
-->

bpapaspyros commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 - feat: Develop KUKA (KR210) Assets and RL Reach Environments (#3)
 - feat: Develop UFactory (xArm6) Assets and RL Reach Environment (#5)
 - fix: Support RLGames and SKRL Reinforcement Learning packages for KR210, xArm6, and UR5E (#7)
+- feat: fwrite a detailed README for learning a policy in Isaac Lab (#9)
```
```suggestion
- feat: write a detailed README for learning a policy in Isaac Lab (#9)
```

bpapaspyros reviewed (commented):
On a first pass this looks really good, just some minor typos.

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+
+After defining the assets, environments, and robot control, creating the necessary training configuration files, and registering the new Gym environment, train a new policy by running the appropriate training script for the Reinforcement Learning Wrappers. Detailed instructions are available [here](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_script ...
+
+Once training is complete and the results meet your expectations, export the model in ONNX format. In RSL-RL, this can be easily achieved using **export_policy_as_onnx**. 
```
Do we need an example here, or where can one find one?

yrh012 replied:
We do, added an example!

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+
+# Next Steps to Execute the Policy on the Robot 
+
+With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it onto real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it in ...
```
"at hand" or "in hand"? I'm not sure

yrh012 replied:
"at hand" feels more natural I think, I'll change that

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+
+# Next Steps to Execute the Policy on the Robot 
+
+With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it onto real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it in ...
```
```suggestion
With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it on real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it into any complex application you develop in AICA Studio.
```

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+
+# Next Steps to Execute the Policy on the Robot 
+
+With the ONNX policy in hand, AICA provides an SDK that enables you to take your simulation-trained policy and deploy it onto real hardware. This [README](https://github.com/aica-technology/dynamic-components/tree/main/source/advanced_components/rl_policy_components) offers a step-by-step guide for deploying a learned policy from Isaac Lab Omniverse to various robotic brands and integrating it in ...
```
The whole document is written in 3rd person, only the last paragraphs use 2nd person ("your expectations", "you develop"). We might be able to keep it 3rd person no?

yrh012 replied:
Indeed!

yrh012 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+Predefined assets configurations are located in the following directory in Isaac Lab Repo:  
+`source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets`.  
+
+This directory contains a range of manipulator robots, including the Franka Panda, Universal Robot UR5E and UR10, Kinova, uFactory, and Kuka. To define a new asset, an asset configuration file must be created within the predefined directory. This file should reference a corresponding USD file. For detailed instructions on importing a new robot not included in the predefined directory, refer to [I ...
```
agreed, I will add the explicit name 👍

yrh012 commented on aica_docs/set_up_reinforcement_learning.md:
````diff
@@ -0,0 +1,373 @@
+Here is an example of defining an articulation configuration to set up an asset in a Reinforcement Learning environment:
+
+```python
+UR10_CFG = ArticulationCfg(
````
Yep, my reasoning here is that I take snippets of code and explain it rather than a full script as the full script are in the repo. Additionally, I think it's better if I link to the original file for completeness.

domire8 replied:
You're not linking to the file though, you could to that on "Here" in the above sentence

yrh012 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+`source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/reach`
+
+### Step 1: Create a New Folder  
+Begin by creating a new folder under the predefined directory. The folder name should typically correspond to the RL actor (in this case, the robot being used).  
```
aha interesting, the actor is the Neural Network that acts upon the Robot, I will re-phrase 👍

yrh012 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+
+Before training a new policy, begin by cloning the AICA fork of [Isaac Lab](https://github.com/aica-technology/isaac-lab). 
+
+Next, refer to the [Isaac Lab Installation](https://isaac-sim.github.io/IsaacLab/main/source/setup/installation/pip_installation.html) guide for comprehensive instructions on installation and developer setup.
```
Indeed, I added more of high level explanation and a verification step for the users before continuing.

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
````diff
@@ -0,0 +1,443 @@
+        scale=1.0,
+    )
+```
+Below is a revised, more fluent version:
+
+---
````
```suggestion
```

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,443 @@
+
+After defining the assets, environments, and robot control, creating the necessary training configuration files, and registering the new Gym environment, train a new policy by running the appropriate training script for the Reinforcement Learning Wrappers. Detailed instructions are available [here](https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_script ...
+
+Below is a cleaned-up, more fluent version:
+
+---
```
```suggestion
```

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
````diff
@@ -0,0 +1,443 @@
+    policy: PolicyCfg = PolicyCfg()
+```
+
+Each observation term specifies how a particular piece of data, such as joint positions or end-effector orientation, is retrieved, optionally corrupted with noise (for dynamic randomization), and appended to the RL agent’s input vector. The `func` parameter identifies a callback function that supplies these values, while the order of terms in the configuration dictates their arrangement in the fi ...
````
```suggestion
Each observation term specifies how a particular piece of data, such as joint positions or end-effector orientation, is 1) retrieved, optionally corrupted with noise (for dynamic randomization), and 2) appended to the RL agent’s input vector. The `func` parameter identifies a callback function that supplies these values, while the order of terms in the configuration dictates their arrangement in the final state input.
```

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
````diff
@@ -0,0 +1,443 @@
+    )
+```
+
+The **ObservationsCfg** class specifies the observations passed to the RL agent in the form of a state vector, which can include information such as joint positions, joint velocities, and end-effector pose for robotic arms.
````
```suggestion
The **ObservationsCfg** class specifies the observations passed to the RL actor in the form of a state vector, which can include information such as joint positions, joint velocities, and end-effector pose for robotic arms.
```

We need to commit to either actor or agent throughout the text

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,443 @@
+
+Before training a new policy, begin by cloning the AICA fork of [Isaac Lab](https://github.com/aica-technology/isaac-lab).
+
+Next, refer to the [Isaac Lab Installation](https://isaac-sim.github.io/IsaacLab/main/source/setup/installation/pip_installation.html) guide for detailed steps on installation and developer setup. Isaac Lab is built on top of Isaac Sim, so first follow the Isaac Sim installation procedure (which can be done through pip). Afterward, install Isaac Lab by adding the necessary development dependencie ...
```
```suggestion
Next, refer to the [Isaac Lab Installation](https://isaac-sim.github.io/IsaacLab/main/source/setup/installation/pip_installation.html) guide for detailed steps on installation and developer setup. Isaac Lab is built on top of Isaac Sim, so first follow the Isaac Sim installation procedure (which can be done through pip). Afterwards, install Isaac Lab by adding the necessary development dependencies and running the installation script in the cloned repository (`./isaaclab.sh --install`).
```

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,443 @@
+Predefined assets configurations are located in the following directory in the Isaac Lab repository:  
+`source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets`.  
+
+This directory contains a range of manipulator robots, including the Franka Emika Panda, Universal Robot UR5e and UR10, Kinova JACO2 (7-Dof), JACO2 (6-Dof) and Gen3 (7-Dof), uFactory xArm 6, and Kuka KR210. To define a new asset, an asset configuration file must be created within [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets). This file should reference a co ...
```
```suggestion
This directory contains a range of manipulator robots, including the Franka Emika Panda, Universal Robot UR5e and UR10, Kinova JACO2, JACO2 and Gen3, uFactory xArm 6, and Kuka KR210. To define a new asset, an asset configuration file must be created within the [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets). This file should reference a corresponding USD file. For detailed instructions on importing a new robot not included in the [lab_assets directory](source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets), refer to [Importing a New Asset](https://isaac-sim.github.io/IsaacLab/main/source/how-to/import_new_asset.html).
```

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,443 @@
+
+## Simulation Environments
+
+Simulation environments are virtual environments where a RL agent takes actions, observes states, and receives rewards. In Isaac Lab, there are two types of environments:
```
```suggestion
Simulation environments are virtual environments with which a RL actor interacts to learn a policy. In Isaac Lab, there are two types of environments:
```

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
````diff
@@ -0,0 +1,443 @@
+    events: EventCfg = EventCfg()
+    curriculum: CurriculumCfg = CurriculumCfg()
+```
+As shown in the configuration above, **ReachEnvCfg** inherits from **ManagerBasedRLEnvCfg** and sets up the various managers that define the RL environment. The **ReachSceneCfg** specify the assets in the environment such as the ground, robot, and lights as demonstrated in the snippet below:
````
```suggestion
As shown in the configuration above, **ReachEnvCfg** inherits from **ManagerBasedRLEnvCfg** and sets up the various managers that define the environment. The **ReachSceneCfg** specify the assets in the environment such as the ground, robot, and lights as demonstrated in the snippet below:
```

domire8 replied:
Inline code blocks instead of bold (across the whole file)

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,373 @@
+
+As shown in the example articulation configuration, the relevant USD file is stored in Isaac Nucleus and can be accessed by importing `ISAACLAB_NUCLEUS_DIR` from `omni.isaac.lab.utils.assets` in Isaac Lab.
+
+Beyond the default assets, AICA has curated a list of additional resources not included in the default Isaac Sim folder on Isaac Nucleus, such as the uFactory xArm 6 and KUKA KR210 robots, which can be made available upon request.
```
Ok makes sense but seems somewhat weird in this document. I feel we should
- either also make the asset configuration available on request
- or find a way to make clear that even though you said a few lines above that these configurations are there, its not usable without USD file

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,443 @@
+1. [Manager-Based RL Environment](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_manager_rl_env.html)  
+2. [Direct RL Environment](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_direct_rl_env.html)
+
+Manager-Based RL Environments simplify simulation complexity by using managers that handle tasks such as fetching, updating, and setting the data required by the RL agent (for example, constructing the actor state from observations). These managers are **InteractiveScene**, **ActionManager**, **ObservationManager**, **RewardManager**, **CurriculumManager**, and **EventManager**.  In contrast, Dir ...
```
```suggestion
Manager-Based RL Environments simplify simulation complexity by using managers that handle tasks such as fetching, updating, and setting the data required by the RL agent (for example, constructing the actor state from observations). These managers are `InteractiveScene`, **ActionManager**, **ObservationManager**, **RewardManager**, **CurriculumManager**, and **EventManager**.  In contrast, Direct-Based RL Environments offer greater flexibility by requiring users to define observations, actions, and rewards directly within the task script.
```
If those are all actual classes, then please with `code block`

domire8 commented:
I propose that you work through these reviews, and then apply the format as I've suggested up top. It's easier to make the reviews when the lines are max 120 characters long

yrh012 commented:
> I propose that you work through these reviews, and then apply the format as I've suggested up top. It's easier to make the reviews when the lines are max 120 characters long. After that we'll read again

Sounds good, I want to keep it to the end to avoid (outated) Review Tag, I'll apply it directly after the above changes.

domire8 commented:
> I want to keep it to the end to avoid (outated) Review Tag

Also a good consideration, we can also do it like that.

yrh012 commented:
I have applied the review requests, updated the linking of files to relative links, updated the bold and code blocks for various entries and formatted using prettier and 120 line-length. Looking forward for your feedback !

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,563 @@
+recurrent actor-critic models Below are example snippets showing how to export both recurrent and feed-forward
+actor-critic models:
+
+`Recurrent model (Memory Based Models)`:
```
```suggestion
### Recurrent model (Memory Based Models):
```

domire8 commented on aica_docs/set_up_reinforcement_learning.md:
````diff
@@ -0,0 +1,563 @@
+)
+```
+
+`Feed-forward model (Common Case)`:
````
```suggestion
### Feed-forward model (Common Case):
```

domire8 reviewed (approved):
Excellent! works for me, thanks for taking into account all the feedback as always

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,563 @@
+vector. The `func` parameter identifies a callback function that supplies these values, while the order of terms in the
+configuration dictates their arrangement in the final state input.
+
+The `ActionsCfg` class defines how the outputs of the Neural Network are translated into actions applied in the
```
```suggestion
The `ActionsCfg` class defines how the outputs of the neural network are translated into actions applied in the
```

I don't see a reason to capitalize `Neural Network` (or `Reinforcement Learning` for that matter) throughout the text, except for when it's in titles which ok given the convention you are using.

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,563 @@
+libraries, translating environment data into the appropriate formats for each library's functions. The three supported
+libraries are:
+
+1. `RSL-RL`: Developed by the Robotic Systems Lab at ETH Zürich, RSL-RL is a fast and straightforward implementation of
```
```suggestion
1. `RSL-RL`: Developed by the Robotic Systems Lab (RSL) at ETH Zürich, RSL-RL is a fast and straightforward implementation of
```

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,563 @@
+### Step 1: Create a New Folder
+
+Begin by creating a new folder. Its location depends on the simulation environment and the specific task being
+performed. For example, for a Manager-Based RL robotic arm manipulation reach task, an appropriate location for that
```
```suggestion
performed. For example, for a Manager-Based RL robotic arm manipulation reach task, an appropriate location for that
```

Here, for example, I'm ok(ish) with capitalizing `Manager-Based RL` only because NVidia is consistently going by this convention.

bpapaspyros reviewed (approved):
Some more comments on my end to consider before merging, but other than that I think it's looking really good

bpapaspyros commented:
for reference, we have plenty more capitalized `Neural Network(s)` instances. Perhaps a case sensitive search will lead you to those instances quickly (some of them should even be replaced by `NN`)

yrh012 commented:
> for reference, we have plenty more capitalized `Neural Network(s)` instances. Perhaps a case sensitive search will lead you to those instances quickly (some of them should even be replaced by `NN`)

Yep indeed, I am doing that now

yrh012 commented:
I decapitalized neural networks and reinforcement learning through out the document @bpapaspyros

domire8 commented:
Feels like a decapitation of NN and RL

yrh012 commented:
> Feels like a decapitation of NN and RL

kinda yes 🤣 felt like lower letter and no abbreviations is more explicit and no special treatment for those words, what do you think?

domire8 commented:
I was just making a bad joke, all good for me :smile:

bpapaspyros commented on aica_docs/set_up_reinforcement_learning.md:
```diff
@@ -0,0 +1,564 @@
+# Reinforcement Learning with Isaac Lab
+
+This document provides a step-by-step guide for training a reinforcement learning (RL) policy based on neural networks
```
```suggestion
This document provides a step-by-step guide for training a Reinforcement Learning (RL) policy based on neural networks
```

first time the abbreviation is defined it needs to be capitalized
